src/lib/support/Pool.h

```diff
@@ -160,4 +167,107 @@ class BitMapObjectPool : public StaticAllocatorBitmap
     } mData;
 };
 
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+
+/**
+ *  @brief
+ *      A class template used for allocating Object subclass objects from an ObjectArena<> template union.
+ *
+ *  @tparam     T   a class to be allocated from the arena.
+ */
+template <class T>
+class HeapObjectPool
+{
+public:
+    template <typename... Args>
+    T * CreateObject(Args &&... args)
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        T * object = new T(std::forward<Args>(args)...);
+        if (object == nullptr)
+        {
+            return nullptr;
+        }
+
+        mObjects.insert(object);
+        return object;
+    }
+
+    void ReleaseObject(T * object)
+    {
+        if (object == nullptr)
+            return;
+
+        std::lock_guard<std::mutex> lock(mutex);
```

Comment:
Similarly, do we need to lock around the object deletion, or should we add a scope so the lock is released before `delete object`, with its possible side-effects, happens?

Reply:
Ditto